packages/blocks/src/types.ts

```diff
@@ -0,0 +1,976 @@
+	 */
+	typography?: Partial< TypographyProps > | undefined;
+
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+//
+// Attributes
+// ----------------------------------------------------------------------------
+
+export namespace AttributeSource {
+	export type Attribute = {
+		source: 'attribute';
+		attribute: string;
+		selector?: string | undefined;
+	} & (
+		| {
+				type: 'boolean';
+				default?: boolean | undefined;
+		  }
+		| {
+				type: 'number';
+				default?: number | undefined;
+		  }
+		| {
+				type: 'string';
+				default?: string | undefined;
+		  }
+	 );
+
+	export interface Children {
+		source: 'children';
+		type: 'array';
+		selector?: string | undefined;
+	}
+
+	export interface HTML {
+		source: 'html';
+		type: 'string';
+		multiline?: 'li' | 'p' | undefined;
+		selector?: string | undefined;
+		default?: string | undefined;
+	}
+
+	export interface Meta {
+		source: 'meta';
+		type: 'string';
+		meta: string;
+		default?: string | undefined;
+	}
+
+	export interface Query< T > {
+		source: 'query';
+		type: 'array';
+		selector: string;
+		query: {
+			[ k in keyof T ]: WPBlockAttribute<
+				T[ k ] extends Array< infer U > ? U : T[ k ]
+			>;
+		};
+		default?: any[] | undefined;
+	}
+
+	export interface Text {
+		source: 'text';
+		type: 'string';
+		selector?: string | undefined;
+		default?: string | undefined;
+	}
+
+	export type None =
+		| ( {
+				source?: never | undefined;
+		  } & (
+				| {
+						type: 'array';
+						default?: any[] | undefined;
+				  }
+				| {
+						type: 'object';
+						default?: object | undefined;
+				  }
+				| {
+						type: 'boolean';
+						default?: boolean | undefined;
+				  }
+				| {
+						type: 'number';
+						default?: number | undefined;
+				  }
+				| {
+						type: 'string';
+						default?: string | undefined;
+				  }
+		   ) )
+		| 'array'
+		| 'object'
+		| 'boolean'
+		| 'number'
+		| 'string';
+}
+
+export type WPBlockAttribute< T > =
+	| AttributeSource.Attribute
+	| AttributeSource.Children
+	| AttributeSource.HTML
+	| AttributeSource.Meta
+	| AttributeSource.Query< T >
+	| AttributeSource.Text
+	| AttributeSource.None
+	| { [ key: string ]: any }; // Permit untyped/undocumented property access.
+
+//
+// Transforms
+// ----------------------------------------------------------------------------
+
+export type TransformRawSchema = {
+	[ k in keyof HTMLElementTagNameMap | '#text' ]?: {
+		attributes?: string[] | undefined;
+		require?: Array< keyof HTMLElementTagNameMap > | undefined;
+		classes?: Array< string | RegExp > | undefined;
+		children?: TransformRawSchema | undefined;
+		[ key: string ]: any; // Permit untyped/undocumented property access.
+	};
+};
+
+export interface TransformBlock< T extends Record< string, any > > {
+	type: 'block';
+	priority?: number | undefined;
+	blocks: string[];
+	isMatch?( attributes: T, block: string | string[] ): boolean;
+	isMultiBlock?: boolean | undefined;
```

Comment:
`__experimentalConvert` can also be present in multi-block transforms

Reply:
I will check that out.

Reply:
Can it be present in all the different types of transforms?

I'll add a `BaseTransform` that has all the shared properties.

Reply:
On DefinitelyTyped we're staying away from publishing experimental and unstable APIs. I guess to keep the compiler happy here that may not be possible, but maybe we can avoid adding them until they become essential?

Or maybe we can find some way to let folks opt-in or to expose them inside Gutenberg without exposing them to eternal consumers. Maybe some declaration merging tricks where the experimental/unstable types are exposed in a different module? 🤔

Reply:
@sirreal I have used `@microsoft/api-extractor` before to rollup declarations and it has the ability to set different release levels. It would be a whole other conversation. But there are developed ways to do exactly what you suggest.

Reply:
@sirreal if it's __experimental and merged to WordPress core then it's really public and could use a public type.